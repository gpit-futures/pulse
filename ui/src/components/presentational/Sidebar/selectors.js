import { createSelector } from 'reselect';

const sidebarVisibilitySelector = ({ isSidebarVisible }) => isSidebarVisible;
const patientsSummarySelector = ({ patientsSummaries }) => patientsSummaries;

const sidebarSelector = createSelector(
  sidebarVisibilitySelector,
  isSidebarVisible => ({ isSidebarVisible })
);

const patientsSummariesSelector = createSelector(
  patientsSummarySelector,
  patientsSummaries => ({ patientsSummaries })
);

export { sidebarSelector, patientsSummariesSelector };
