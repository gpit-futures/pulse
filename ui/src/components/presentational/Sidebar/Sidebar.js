import React, { PureComponent } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom'
import _ from 'lodash/fp';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { sidebarConfig } from '../../../plugins.config';
import { sidebarSelector, patientsSummariesSelector } from './selectors';
import { setSidebarVisibility } from '../../../ducks/set-sidebar-visibility';

const mapDispatchToProps = dispatch => ({ actions: bindActionCreators({ setSidebarVisibility }, dispatch) });

@connect(sidebarSelector, mapDispatchToProps)
@connect(patientsSummariesSelector, mapDispatchToProps)
export default class Sidebar extends PureComponent {
  static propTypes = {
    activeLink: PropTypes.string,
    userId: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  };

  /* istanbul ignore next */
  componentWillMount() {
    window.addEventListener('resize', () => {
      this.setPositionForSidebar()
    });
    window.addEventListener('scroll', () => {
      this.setPositionForSidebar()
    });
    window.addEventListener('orientationchange', () => {
      this.setPositionForSidebar()
    });
    if (_.isEmpty(this.props.patientsSummaries)) {
      this.hideSidebarOnMobile();
    }
  }

  /* istanbul ignore next */
  componentDidMount() {
    this.setPositionForSidebar();
  }

  /* istanbul ignore next */
  componentWillReceiveProps(nextProps) {
    if (nextProps.patientsSummaries.length !== 0) {
      this.setPositionForSidebar();
    }
  }

  /* istanbul ignore next */
  setPositionForSidebar() {
    const page = document;
    const headerHeight = page.getElementsByClassName('header')[0] ? page.getElementsByClassName('header')[0].offsetHeight : 0;
    const footerHeight = page.getElementsByClassName('footer')[0] ? page.getElementsByClassName('footer')[0].offsetHeight : 0;
    const sidebar = page.getElementsByClassName('sidebar')[0];
    const sidebarUnderlay = page.getElementsByClassName('sidebar-underlay')[0];

    if (sidebar !== undefined) {
      const scrollPageTop = Math.max(document.documentElement.scrollTop, document.body.scrollTop);
      let sidebarTop = headerHeight - scrollPageTop;

      sidebarTop = sidebarTop > 0 ? sidebarTop : 0;

      sidebar.style.top = `${sidebarTop}px`;
      sidebarUnderlay.style.top = `${sidebarTop}px`;

      if (window.innerWidth < 768) {
        sidebar.style.bottom = 0;
        sidebarUnderlay.style.bottom = 0;
      } else {
        sidebar.style.bottom = `${footerHeight}px`;
        sidebarUnderlay.style.bottom = `${footerHeight}px`;
      }
    }
  }

  toggleSidebarVisibility = /* istanbul ignore next */ () => {
    const { actions, isSidebarVisible } = this.props;
    if (window.innerWidth < 768) {
      actions.setSidebarVisibility(!isSidebarVisible);
    }
  };

  hideSidebarOnMobile = () => {
    const { actions } = this.props;
    /* istanbul ignore next */
    if (window.innerWidth < 768) {
      /* istanbul ignore next */
      actions.setSidebarVisibility(false);
    }
  };

  render() {
    const { activeLink, userId } = this.props;
    return (
      <div>
        <div className="sidebar-underlay showSidebar" />
        <div className="sidebar showSidebar" role="navigation">
          <div className="sidebar-nav">
            <div>
              <ul className="sidebar-nav-list">
                {sidebarConfig.map((item, index) => (item.isVisible ? <li className="sidebar-nav-item" key={index}>
                  <Link className={classNames('sidebar-nav-link', { active: activeLink === item.key })} to={`/patients/${userId}${item.pathToTransition}`} onClick={this.toggleSidebarVisibility}>{item.name}</Link>
                </li> : null))
                }
              </ul>
            </div>
          </div>
        </div>
      </div>
    )
  }
}
