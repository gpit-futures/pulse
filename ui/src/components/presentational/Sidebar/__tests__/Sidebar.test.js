import React from 'react';
import configureStore from 'redux-mock-store';
import sinon from 'sinon'
import { StaticRouter } from 'react-router'
import Enzyme, { shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';

import Sidebar from '../Sidebar';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();
const dispatch = sinon.spy();
const userId = 9999999000;
const activeLink = 'patients-summary';
const context = {};
const testUnnecessaryProp = 'test';

describe('Component <Sidebar />', () => {
  it('should renders with props correctly', () => {
    const sidebar = shallow(
      <Sidebar
        dispatch={dispatch}
        store={mockStore({ runtime: {} })}
        userId={userId}
        activeLink={activeLink}
        testUnnecessaryProp={testUnnecessaryProp}
      />);

    const testUnnecessaryProps = sidebar.instance().props.testUnnecessaryProp;

    // init rendering with default props
    expect(sidebar).toMatchSnapshot();
    expect(sidebar.prop('activeLink')).toEqual('patients-summary');
    expect(sidebar.prop('userId')).toEqual(9999999000);
    expect(testUnnecessaryProps).toEqual(testUnnecessaryProp);

    // simulate transition on the another route and change activeLink
    sidebar.setProps({ userId: 9999999024, activeLink: 'allergies' });
    expect(sidebar).toMatchSnapshot();
    expect(sidebar.prop('activeLink')).toEqual('allergies');
    expect(sidebar.prop('userId')).toEqual(9999999024);
  });

  it('should renders when it is visible', () => {
    const sidebarConfig = [
      { key: 'patients-summary', pathToTransition: '/patients-summary', name: 'Patient Summary', isVisible: true },
      { key: 'diagnoses', name: 'Problems / Diagnosis', isVisible: true },
    ];

    const tree = render(
      <StaticRouter location="someLocation" context={context}>
        <Sidebar
          dispatch={dispatch}
          store={mockStore({ runtime: {} })}
          userId={userId}
          activeLink={activeLink}
          isSidebarVisible
        />
      </StaticRouter>);
    expect(tree).toMatchSnapshot();
    expect(tree.find('.active')).toHaveLength(1);
  });
});
