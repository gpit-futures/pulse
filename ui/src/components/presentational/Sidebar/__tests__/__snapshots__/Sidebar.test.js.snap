// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <Sidebar /> should renders when it is visible 1`] = `
<div>
  <div
    class="sidebar-underlay showSidebar"
  />
  <div
    class="sidebar showSidebar"
    role="navigation"
  >
    <div
      class="sidebar-nav"
    >
      <div>
        <ul
          class="sidebar-nav-list"
        >
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link active"
              href="/patients/9999999000/patients-summary"
            >
              Patient Summary
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/diagnoses"
            >
              Problems / Diagnosis
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/medications"
            >
              Medications
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/allergies"
            >
              Allergies
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/contacts"
            >
              Contacts
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/events"
            >
              Events
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/documents"
            >
              Documents
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/orders"
            >
              Orders
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/results"
            >
              Test Results
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/procedures"
            >
              Procedures
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/clinicalNotes"
            >
              Clinical Notes
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/clinicalStatements"
            >
              Clinical Statements
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/personalNotes"
            >
              Personal Notes
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/vaccinations"
            >
              Vaccinations
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/vitals"
            >
              Vitals - News
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/images"
            >
              Images
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/drawings"
            >
              Drawings
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/referrals"
            >
              Referrals
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/mdt"
            >
              MDT
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/transfer-of-care"
            >
              Transfers of Care
            </a>
          </li>
          <li
            class="sidebar-nav-item"
          >
            <a
              class="sidebar-nav-link"
              href="/patients/9999999000/proms"
            >
              PROMs
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`Component <Sidebar /> should renders with props correctly 1`] = `
<Connect(Sidebar)
  actions={
    Object {
      "setSidebarVisibility": [Function],
    }
  }
  activeLink="patients-summary"
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  testUnnecessaryProp="test"
  userId={9999999000}
/>
`;

exports[`Component <Sidebar /> should renders with props correctly 2`] = `
<Connect(Sidebar)
  actions={
    Object {
      "setSidebarVisibility": [Function],
    }
  }
  activeLink="allergies"
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  testUnnecessaryProp="test"
  userId={9999999024}
/>
`;
